Prometheus metrics

ath88 opened the pull request.

ath88 committed:
Express doesn't take _any_ arguments at all on initial call

ath88 committed:
Pass config to setup

ath88 committed:
Run example with nodemon instead of node

ath88 committed:
Whitespace fix

ath88 committed:
Add prometheus metrics options

ath88 committed:
Add prometheus metrics port config to example

ath88 commented:
dont review this. express-prometheus-middleware is deprecated

ath88 commented:
https://github.com/joao-fontenele/express-prometheus-middleware/issues/71

ath88 commented:
Hang on while this hopefully resolves quickly.

https://github.com/jochen-schweizer/express-prom-bundle/pull/101

ath88 committed:
Change to use express-prom-bundle

ath88 commented:
@hypesystem Ready for review

hypesystem commented on package.json:
```diff
@@ -4,12 +4,13 @@
     "express": "^4.18.1",
     "express-async-handler": "^1.2.0",
-    "express-prometheus-middleware": "^1.2.0",
+    "express-prom-bundle": "^6.5.0",
```
I see only express-prom-bundle, but also a dexpress-prom-bundle in the package-lock file? We might need a lockfile cleanup

ath88 replied:
You're right. [326c423](https://github.com/derangeddk/dexpress/pull/4/commits/326c42369f132805271a57a9e8e89675eb371791)

hypesystem commented on index.js:
```diff
@@ -1,12 +1,12 @@
 
-async function proxy() {
-    const app = express.apply(express, arguments);
+async function proxy(config = { }) {
```
Syntax: prefer `{}` over `{ }` here :smile:

ath88 replied:
I prefer `{ }`. :D We could put this into a linter-rule so we have a forum for more general discussions and agreements on this. See https://gitlab.deranged.dk/deranged/eslint-plugin

hypesystem replied:
😢

hypesystem commented on setUpApplicationDefaults.js:
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
+            delete metricsConfig.port;
+        }
+        app.use(promBundle({
+          metricsApp: app.metricsApp,
```
What is metricsApp in this case, if `port` is falsy?

Does it make sens to listen to the app right off the bat, too? I think that seems a little over-eager to me. I would prefer changing the main `app.listen` to first start the metricsApp listening instead? That would align closer to my expected behavior.

ath88 replied:
It's undefined, which is fine for promBundle.

I've changed startup to happen during `app.listen` instead. [a059624](https://github.com/derangeddk/dexpress/pull/4/commits/a059624936079571ab93999a11d532cbf6f99c21)

hypesystem commented on setUpApplicationDefaults.js:
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
+          autoregister: false,
+          includeMethod: true,
+          includePath: true,
+          promClient: { collectDefaultMetrics: { } }
```
Now that we're talking about changes, I think supporting more options in `createApplication` would be nice - maybe we can expose these options somehow?

ath88 replied:
[9470cba](https://github.com/derangeddk/dexpress/pull/4/commits/9470cbae86b1fdc7cae4215d7d15b2223b067690)

hypesystem commented on setUpApplicationDefaults.js:
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
+          promClient: { collectDefaultMetrics: { } }
+        }));
+    }
+    // TODO we need to terminate app.metricsApp too somehow on shutdown
```
Just like I suggested with overwriting app.listen, I think we should do the same for server.close, probably...

ath88 replied:
I see a problem in trying to extend server.close. It's a deep native thing, so making it magic feels weird to me.

I don't really like patching native code with too much magic, so we need to take a stand here. Same problem with app.listen, which is not a promise. We could make it a promise, but we would be breaking expectations here.

hypesystem replied:
Agreed -- but we need some way to ensure shutdown happens properly, right? What's the current solution to this?

hypesystem replied:
Agreed -- but we need some way to ensure shutdown happens properly, right? What's the current solution to this?

ath88 replied:
[7885c55](https://github.com/derangeddk/dexpress/pull/4/commits/7885c55dea99045736851a9746fb817ec7c7a527)

hypesystem reviewed (changes requested):
Some significant questions about behavior vs expected behavior.

I generally love the approach, I just want to make sure we hit a good/easy to use solution here.

ath88 committed:
Lock-file cleanup

ath88 committed:
Auto-register prometheus metrics on primary app if no other port is set

ath88 committed:
Start metrics app on listen

Extending a non-promise method with callbacks feels old school, but bad

ath88 committed:
Expose access to promClient configuration

ath88 committed:
Move dependency

ath88 committed:
Ensure proper application termination

ath88 commented:
@hypesystem Should be fixed now :)

hypesystem commented on setUpApplicationDefaults.js:
```diff
@@ -64,7 +69,25 @@ export default async (app) => {
+          },
+        }));
+    }
+    // TODO we need to terminate app.metricsApp too somehow on shutdown
```
This is now handled, comment should be removed!

ath88 replied:
Done

hypesystem commented on setUpApplicationDefaults.js:
```diff
@@ -64,7 +69,25 @@ export default async (app) => {
-    app.use(prometheus());
+
+    // Add prometheus monitoring metrics
+    if (config.prometheusMetrics) {
```
Reasonable defaults: is it on purpose that we default to not having a prometheus metrics app at all? That may be reasonable enough, I haven't thought it through...

But I want the general approach of this lib to be "don't worry, we do the good thing per default, but you can config it or toggle it off" - I don't know if we're hitting that here with default-off.

ath88 replied:
Agree. [5fc3eae](https://github.com/derangeddk/dexpress/pull/4/commits/5fc3eae4934ab82c51e608e42f021e368b014a15)

hypesystem reviewed (approved):
Looks good! Generally approved, some few last questions, but not blocking for a merge and publish :smile:

ath88 committed:
remove void comment

ath88 committed:
Reorder for clarity

ath88 committed:
Changes default to add metrics to existing app

I would prefer to default to another port, but that requires that we can 
find one that is unused. For now its probably better to just mount it on 
the existing express app.

It can be disabled with `{ prometheusMetrics: { enabled: false } }`.
